File: 프로그래머스/1/140108. 문자열 나누기/문자열 나누기.js
function solution(s) {
    const splited = [];
    
    decompose(s);
    
    return splited.length;
    
    function decompose(str) {
        if (str.length < 1) return;
        
        const first = str[0];
        const count = [0, 0];
        
        for (let i = 0; i < str.length; i++) {
            const char = str[i];
            
            if (char === first) {
                count[0]++;
            } else {
                count[1]++;
            }
            
            if (count[0] === count[1]) {
                splited.push(str.slice(0, i + 1));
                decompose(str.slice(i + 1));
                
                return;
            }
        }
        
        splited.push(str);
    }
}

File: 프로그래머스/1/140108. 문자열 나누기/문자열 나누기.py
def solution(s):
    def split_string(string):
        result = string[:]
        head = result[0]
        count = [1, 0]
        
        for i in range(1, len(result)):
            if (head == result[i]):
                count[0] += 1
            else:
                count[1] += 1
                
            if count[0] == count[1]:
                return [result[:i], result[i+1:]]
            
        return [result, '']

    count = 0
    
    while len(s) > 0:
        _, ns = split_string(s)
        count += 1
        s = ns
        
    return count
